### PR TITLE
Add System Catalog Views for Partition Metadata

### PR DESCRIPTION
**Pull request by sumitj824**
### Description

This commit introduces support for the following system views:

1.  sys.partitions: To provide information about the individual partitions present in the database for both partitioned and non-partitioned indexes/tables.
2. sys.data_spaces: To provide information about the data spaces utilized by partition schemes.
3. sys.destination_data_spaces: To provide information about the destination data spaces associated with partition schemes.

Furthermore, this commit also updates the definition of the following existing views:

1. sys.indexes: Updates the definition to list HEAP entries for partitioned tables and updates the data_space_id column to the ID of the associated partition scheme.
2. sys.index_columns: Updates the definition to list HEAP entries for partitioned tables and entries for indexes of partitioned tables, which will be used to find the partitioning column.
3. sys.filegroups: Updates the datatype of the name column from SYSNAME to NVARCHAR(128).

These views, along with sys.partition_functions, sys.partition_schemes, sys.partition_parameters, and sys.partition_range_values which was added in 1ad41811a76292d7cdcc4402a080288118d1d233, provide comprehensive information about the partitioning configuration in the database.

Signed-off-by: Sumit Jaiswal [sumiji@amazon.com](mailto:sumiji@amazon.com)

This PR is stacked over https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2702

### Issues Resolved

Task: BABEL-4970

### Test Scenarios Covered ###
* **Use case based -** Yes

* **Boundary conditions -** NA

* **Arbitrary inputs -** NA

* **Negative test cases -** NA

* **Minor version upgrade tests -** Yes

* **Major version upgrade tests -** Yes

* **Performance tests -** NA

* **Tooling impact -** NA

* **Client tests -** Yes

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 1]
```diff
@@ -3219,19 +3297,110 @@ FROM sys.events e
...
+  CAST('OFF' as sys.varchar(3)) as xml_compression_desc
+FROM pg_inherits pgi
+INNER JOIN pg_index cidx on cidx.indexrelid = pgi.inhrelid
+INNER JOIN pg_index pidx on pidx.indexrelid = pgi.inhparent
+INNER JOIN pg_class ctbl on ctbl.oid = cidx.indrelid
+INNER JOIN index_id_map imap on imap.indexrelid = cidx.indexrelid
+INNER JOIN sys.schemas sch on sch.schema_id = ctbl.relnamespace
+WHERE ctbl.relkind = 'r'
+AND ctbl.relispartition
+AND has_schema_privilege(ctbl.relnamespace, 'USAGE');
```
Comment (shah-nirmit) [inline contrib/babelfishpg_tsql/sql/sys_views.sql]: Can we also add indislive filter here?

**Review by shah-nirmit — APPROVED** [review]

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 2]
```diff
@@ -2548,15 +2600,41 @@ WHERE
...
 SELECT 
-  CAST('PRIMARY' as SYSNAME) AS name,
+  CAST('PRIMARY' as sys.NVARCHAR(128)) AS name,
   CAST(1 as INT) AS data_space_id,
   CAST('FG' as sys.bpchar(2)) AS type,
-  CAST('ROWS_FILEGROUP' as NVARCHAR(60)) AS type_desc,
+  CAST('ROWS_FILEGROUP' as sys.NVARCHAR(60)) AS type_desc,
   CAST(1 as sys.BIT) AS is_default,
-  CAST(0 as sys.BIT) AS is_system;
+  CAST(0 as sys.BIT) AS is_system
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/sql/sys_views.sql:2627]: do we really need dummy entry now?

Comment (sumitj824) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:2627]: yes, we will need to entry of primary filegroup which will be used for DDL/Export.
